src/filter.rs

```diff
@@ -132,12 +132,100 @@ pub(crate) fn unfilter(
         }
     }
 
+    // SIMD Within a Register, see "Unpartitioned Operations With Correction Code"
+    // http://aggregate.org/SWAR/over.html
+    fn swar_add_u32(x: u32, y: u32) -> u32 {
+        // Do 7-bit addition so there's no carry over the most significant bit
+        let n = (x & 0x7f7f7f7f) + (y & 0x7f7f7f7f);
+        // 1-bit parity/XOR addition to fill in the missing MSB
+        n ^ (x ^ y) & 0x80808080
+    }
+
+    fn swar_add_u64(x: u64, y: u64) -> u64 {
+        let n = (x & 0x7f7f7f7f7f7f7f7f) + (y & 0x7f7f7f7f7f7f7f7f);
+        n ^ (x ^ y) & 0x8080808080808080
+    }
+
+    // Average of two integers without overflow or casting to a larger bit-width
+    // http://aggregate.org/MAGIC/#Average%20of%20Integers
+    // The base algorithm:
+    //     (x & y) + ((x ^ y) >> 1)
+    fn swar_avg_u32(x: u32, y: u32) -> u32 {
+        // Calculate 4 averages at once masking out the MSB on the right operand
+        // so right-shifted values don't interfere with their neighbors.
+        (x & y) + (((x ^ y) >> 1) & 0x7f7f7f7f)
+    }
+
+    fn swar_avg_u64(x: u64, y: u64) -> u64 {
+        (x & y) + (((x ^ y) >> 1) & 0x7f7f7f7f7f7f7f7f)
+    }
+
     match filter {
         NoFilter => Ok(()),
         Sub => {
             let current = &mut current[..len];
-            for i in bpp..len {
-                current[i] = current[i].wrapping_add(current[i - bpp]);
+            if bpp > len {
+                return Err("Filtering failed: bytes per pixel is greater than length of row");
+            }
```

Comment:
This is a new error path that makes it possible for `Sub` to return an error that would have been `Ok` before.

I'm not sure of the strategy for handling `bpp > len` because `Avg` and `Paeth` already return it. Should it be hoisted out of the match and apply to all filters, or is this fine?

Reply:
Honestly, unfilter probably shouldn't return a Result in the first place. Assertions are fine for code clarity or to inform the optimizer, but the method is only called from a single place with arguments that should have already been validated:

https://github.com/image-rs/image-png/blob/56e3ea6000b825c90e02e5df743bb0479a686294/src/decoder/mod.rs#L777

Like staring at that line, it seems pretty silly to have `require_length` and all the code around validating `previous` and `current` have the same length, when they are both guaranteed to be exactly `rowlen-1` bytes long!

Reply:
Oh, I kept losing track of the caller for `unfilter` with all the context switches between `decoder::mod` and `decoder::filter` so I wasn't sure what data `unfilter` was getting.

The validation of `require_length` is completely unnecessary then. I wasn't sure if the `bpp > len` check was anywhere else but it seems like it should definitely live at least one layer above this function in `next_raw_interlaced_row`.

I've made it so the function no longer returns a result and the error check is done in the caller function.